ipn/ipnlocal/local.go

```diff
@@ -4126,10 +4126,23 @@ func shouldUseOneCGNATRoute(logf logger.Logf, controlKnobs *controlknobs.Knobs,
 //
 // The versionOS is a Tailscale-style version ("iOS", "macOS") and not
 // a runtime.GOOS.
-func dnsConfigForNetmap(nm *netmap.NetworkMap, peers map[tailcfg.NodeID]tailcfg.NodeView, prefs ipn.PrefsView, logf logger.Logf, versionOS string) *dns.Config {
+func dnsConfigForNetmap(nm *netmap.NetworkMap, peers map[tailcfg.NodeID]tailcfg.NodeView, prefs ipn.PrefsView, selfExpired bool, logf logger.Logf, versionOS string) *dns.Config {
 	if nm == nil {
 		return nil
 	}
+
+	// If the current node's key is expired, then we don't program any DNS
+	// configuration into the operating system. This ensures that if the
+	// DNS configuration specifies a DNS server that is only reachable over
+	// Tailscale, we don't break connectivity for the user.
+	//
+	// TODO(andrew-d): this also stops returning anything from quad-100; we
+	// could do the same thing as having "CorpDNS: false" and keep that but
+	// not program the OS?
```

Comment:
I'd rather keep the DNS config and program the OS unless the Tailnet's _Global nameservers_ list contains only DNS servers that are reachable exclusively over Tailscale. And also blocked/returned SERVFAIL for queries covered by the customer's split DNS.

This would help prevent DNS leaks while the node key is expired. If a customer has a preferred public DNS server they trust, there's really no reason not to use it with an expired key. And I believe we shouldn't send the customer's internal DNS names (those covered by their split DNS config) to a random DNS server at a cafe.